spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -804,6 +804,33 @@ impl Utf8String {
     }
 }
 
+// Index
+impl Utf8String {
+    #[inline]
+    #[must_use]
+    pub fn index(&self, needle: &[u8], offset: usize) -> Option<usize> {
+        for byte_pos in self.find_iter(needle) {
+            let char_pos = self[..byte_pos].chars().count();
```

Comment:
this does an unchecked index into a byte slice (ok) and calls `bstr::ByteSlice::chars()` on the result. `chars` replaces invalid UTF-8 byte sequences with the Unicode replacement character `�`. This means all of these cases return incorrect indexes:

- `Utf8String`s which contain invalid UTF-8 byte sequences.
- Needles which contain invalid UTF-8 byte sequences that match a prefix or suffix of a UTF-8 character in the `Utf8String`.

`Utf8String::index` and `Utf8String::rindex` should use the `bstr::decode_utf8` or `bstr::decode_last_utf8` functions to do a "character"-at-a-time decoding. These routines allow properly detecting whether an invalid UTF-8 byte sequence has been encountered and properly juggling the indexes.

See this example in `Utf8String::get_char` for how to use these APIs:

https://github.com/artichoke/artichoke/blob/b49699f43241f24b5781efe7de3f225b16c297ac/spinoso-string/src/enc/utf8/mod.rs#L237-L281

Reply:
Thanks! I updated both `enc::utf8::index` and `enc::utf8::rindex` to use `bstr::decode_utf8` in b1d69c7.

<img width="1457" alt="image" src="https://github.com/artichoke/artichoke/assets/12410942/3be0ad1e-695f-40d8-86f8-e0cdf773c16a">